Fix #549, #494 - Use CacheStorage and Service Worker in Live Dev Server

humphd opened the pull request:
I'm not 100% done this yet, but it's working, so I thought I'd get it up here in case others are interested.  This PR does a number of things:

* Replaces the `BlobUtils` module with `UrlCache`, which picks a URL strategy (Blob URLs or real URLs cached by Service Worker) at run time based on what the browser can support.  I've also converted the code to run async (again!), since we need to interact with `window.caches`, and it's `Promise` based.
* When we serve files from the filesystem via the Service Worker, they'll have a URL that looks something like this: `http://localhost:8000/dist/vfs/7/projects/30/tutorial.html`.  NOTE the `dist/vfs/` addition. We need to create a "virtual root" for the filesystem paths, since we need the files to not exist in Brackets itself.  This means we can't put files in `src/vfs/` or `dist/vfs/`, so I've added that to `.gitignore`.
* I've added a new script to be imported to our `bramble-sw.js` service worker, which will handle serving content from the service worker's cache when needed.
* I've switched the Brackets preference code to use `MemoryStorage` vs. file storage, because I want to avoid a timing issue with the filesystem starting and the Cache being initialized.  There is still one file be written that I don't want `Application Support/brackets.json`, but it's not hurting anything.  We don't use either of these files (although we could, if we ever decide to move all our preferences into Brackets vs. Thimble).
* I have some `// TODO`s scattered through the code still, because I'm not sure about a few things, and I need to test more.  We should make sure these all get dealt with before we move to merge this.

To run this, you have a few options.  First, I have some debug code in `UrlCache.js`:

```
function init(callback) {
+        // Prefer CacheStorage if we have access to it.
+        _provider = 'caches' in window ?
+            new CacheStorageUrlProvider() :
+            new BlobUrlProvider();
+
+        // TODO: for testing, uncomment this, currently forcing blob urls.
+        _provider = new BlobUrlProvider();
```

It will continue to use Blob URLs if you leave it like this.  If you want to test the Service Worker/Cache version, do the following:

* remove that line (or comment out)
* run `npm run build`
* run `grunt swPrecache`
* load `http://localhost:8000/dist/hosted.html`.

You might need to force your Service Worker to unregister and reload on startup.  The easiest way to achieve this is by doing this:

![screen shot 2017-04-18 at 4 34 24 pm](https://cloud.githubusercontent.com/assets/427398/25151807/f51a9b02-2454-11e7-96c1-ba2a6314bbfa.png)

NOTE: I'm not wiping my databases, so the filesystem won't get blown away.

Now when you run the app, you can do things like this:

![screen shot 2017-04-18 at 4 36 25 pm](https://cloud.githubusercontent.com/assets/427398/25151881/336639a2-2455-11e7-9f31-0e45546565a1.png)

Here we are loading an image dynamically with a script:

```js
var img = new Image();
img.onload = function() {
  document.body.appendChild(img);
};
img.onerror = function(err) {
  console.log("error loading img", err);
}
img.src = "thumbs-up-sign.png?" + Date.now();
```

That used to be impossible, because it would need to get rewritten.  You can see that all my files are cached in Cache Storage, and URLs will work as expected, because I inject an appropriate `base` element:

![screen shot 2017-04-18 at 4 38 15 pm](https://cloud.githubusercontent.com/assets/427398/25151947/76395156-2455-11e7-8762-f896fff3e946.png)

So a relative file path, used anywhere in that `window`, will get turned into an absolute URL into our `vfs` cache for the project, and "Just Work".

I'm really please with how easy this was to do without having to destroy all of our other work.  We have to keep supporting both ways, since [not all browsers support this approach yet](https://jakearchibald.github.io/isserviceworkerready/).

Probably this will take until mid- to late-May to get 100% finished, tested, and landed.  But it's 100% going to work.  Amazing!

cvan commented on src/bramble-live-dev-cache-sw.js:
```diff
@@ -0,0 +1,25 @@
+    "use strict";
+
+    // Strip params off URL so it will properly match what's in cache
+    var url = event.request.url.split(/[?#]/)[0];
```
since `URL` has been around since the Service Worker API has been around, you can use `URL` without worrying about poor browser support:

```js
var url = new URL(window.location.href).pathname;
```

secretrobotron commented:
😄

gideonthomas commented on src/bramble-live-dev-cache-sw.js:
```diff
@@ -0,0 +1,25 @@
+ * which is named vfs/project/root.
+ */
+
+self.addEventListener("fetch", function(event) {
```
won't this always be the `window` object?

humphd replied:
No, this is running in the worker, and there will never be a `window`.  Do you mean why am I using `self`? vs. just `addEventListener`?  I could switch, just force of habit.

humphd commented:
@gideonthomas I have some more commits I'll push up for this if you're starting to test, which make it a lot easier to test.  I'll try to push those tonight.

humphd commented:
Pushed a bunch more of this.  One thing I've added is the ability to use `?cacheType=blob` or `?cacheType=cacheStorage` on the URL to force one or the other caching strategy at runtime.  I've also made it possible to work with `?cacheType=cacheStorage` in `src/` so you don't have to build in order to test this.

I'm chasing a bug right now with how live CSS documents work, where editing a CSS file will cause an infinite recursive loop of `updateDoc` event and `reload` calls.  It doesn't happen with blob URLs, I suspect because these change every time there is a save, and I must have logic somewhere along the chain that is looking at the stability of the URL.  I'll get back to this when I'm home from PTO in a week and a half.

@gideonthomas if you do feel like trying this out, maybe you'll find other bugs I'm not aware of yet.  The basics work well for me, but I'm sure there's more trouble lurking like this CSS bug.

humphd commented:
Note to myself to test the project listed in https://github.com/mozilla/thimble.mozilla.org/issues/2036 with the new code, it's a good example of what users expect to be able to do, and can't at the moment.

humphd commented:
Note to myself, when https://github.com/mozilla/brackets/pull/704 lands, I need to deal with BlobURL vs. regular URLs for download.

humphd commented:
I tested that project above that does JS image loading with scripts, and it works well:

![screen shot 2017-06-06 at 3 47 12 pm](https://user-images.githubusercontent.com/427398/26848743-80723eee-4acf-11e7-95a1-f2ca9aa66457.png)

humphd commented:
I rebased this, and rewrote the XHR shim code so it a) matches our other injected scripts; b) does the right thing in Blob vs. Cached URL modes.  I tested both, and they work as intended, with the Cached URL one using the network and native XHR vs. the shim.

humphd commented:
I also dealt with the Download URL stuff from above.

humphd commented:
Removing `WIP` from the title, I think this is pretty much done and ready for testing.  I'll throw it up on my gh-pages branch so people can try it.

humphd commented:
I guess I can't use gh-pages, since it isn't `https`.  However, this is working really well!  Check out running [TODO MVC apps](http://todomvc.com/):  I just did a drag-and-drop of the files/folders into the file tree, and it Just Works!

Here's the Vanilla JS:

![screen shot 2017-06-06 at 6 09 14 pm](https://user-images.githubusercontent.com/427398/26854274-6e54a1e8-4ae3-11e7-99e5-625ed4c567f2.png)

Here's the React version unmodified and running perfectly:

![screen shot 2017-06-06 at 6 11 55 pm](https://user-images.githubusercontent.com/427398/26854331-b00ef5c0-4ae3-11e7-9340-aca8c3070407.png)

cvan commented:
> I guess I can't use gh-pages, since it isn't https

You can enable HTTPS on GitHub Pages from the Settings page (`Enforce HTTPS`). Or, if you have a custom `CNAME`, you can set up with your host.

humphd commented:
Cool, I'll investigate and see if I can get it auto-hosted on gh-pages.

In the meantime, I managed to get the code for [this Mozilla HTML5 game workshop](https://hacks.mozilla.org/2017/04/html5-games-workshop-make-a-platformer-game-with-javascript/) running using this PR, see https://github.com/mozilla/thimble.mozilla.org/issues/1955#issuecomment-306662729:

![screen shot 2017-06-06 at 9 47 00 pm](https://user-images.githubusercontent.com/427398/26858951-81d4a3ac-4b02-11e7-81d8-7d266e4b5020.png)

@gideonthomas, let's find some time soon to review this together, and test.  I'm really pleased with how it's working, and what it enables.

cvan commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -61,11 +61,20 @@ define(function (require, exports, module) {
+
+        // Warn that some network features won't work fully
+        if(warn) {
+            console.info("[Bramble] Your browser doesn't support Service Workers and/or CacheStorage. "  +
```
should this be `console.warn` instead, or is this fine?

humphd replied:
Either is [technically fine](https://dxr.mozilla.org/mozilla-central/source/dom/webidl/Console.webidl#21), but I think you're right that `warn` does what I want more clearly.  I'll update.

flukeout commented:
Testing now

flukeout commented:
Okay awesome, I tested the sample code that you provided with setting the img src dynamically and that worked great. I also tested some code I wrote that uses the web audio API, and loads local sound files with an ``XMLHttpRequest`` and guess what, that worked without any problems too!

![image](https://user-images.githubusercontent.com/25212/27055088-1c42faa0-4f78-11e7-8e43-245df2f30a87.png)

flukeout commented:
**Tested**

* [x] Making an XMLHttpRequest to a local audio file
* [x] Setting an image source dynamically via javascript to a local file
* ... what are some other good test cases?

@Pomax @gideonthomas

humphd commented:
Another thing to test is that I didn't break the existing`Blob URL` code, since we'll still need that in browsers that don't support Service Worker/Cache Storage (e.g., Safari, IE11, Edge).

NOTE: testing in those browsers will mean this fancy JS stuff you're doing will break; but that's how it is today as well.  So I guess testing with HTML, CSS, images, etc.

flukeout commented:
Not sure if it's related but, getting some memory leaks in the console.

![image](https://user-images.githubusercontent.com/25212/27094491-12440f10-501f-11e7-9645-34630befe22e.png)

Steps to reproduce...

* Create a new project
* Click the Reload button in the preview a lot and quickly
* After about 5-10 clicks, this error comes up
* Once it comes up, it happens every time you refresh, no matter how quickly

humphd commented:
Can you expand the stack on those so I can see what's actually being called?

flukeout commented:
Links to other pages within the project while editing open in a new browser tab, like this...

This link...
![image](https://user-images.githubusercontent.com/25212/27094733-f36b75e6-501f-11e7-8ecf-e35c7ea8b4cd.png)

Opens up a new tab...
![image](https://user-images.githubusercontent.com/25212/27094723-e83ac7a8-501f-11e7-98b0-4c909d196def.png)

humphd commented:
@flukeout Ah, nice catch!  That's the link manager being fooled by the `http://` portion of the new URLs.  I'll fix.

flukeout commented:
@humphd memory leak expanded...

![image](https://user-images.githubusercontent.com/25212/27095039-153cbd14-5021-11e7-9734-a7def589d13e.png)

![image](https://user-images.githubusercontent.com/25212/27095052-1f7ba4d4-5021-11e7-95d9-3caeb0086b7f.png)

flukeout commented:
Some more errors when refreshing the preview hyper-fast

![image](https://user-images.githubusercontent.com/25212/27095119-74066354-5021-11e7-8de9-09d44e1b10af.png)

This one seems to relate to maintaining the previous scroll position.

humphd commented:
@flukeout I can't reproduce your "links open in new tab."  Are you linking to the same page within itself?

I tried making a link to `tutorial.html` in the `index.html` and it worked as expected (i.e., opened it in the editor, changed the preview).

flukeout commented:
@humphd I can't get it to happen now in a new project I just made - it was def happening in a previous one so I'll keep an eye on it.

flukeout commented:
Getting console errors when "auto-update" is disabled and I add a new element to the page and then add content to it...

![image](https://user-images.githubusercontent.com/25212/27095606-2bc637fc-5023-11e7-9a3b-e68c1eae2d38.png)

humphd commented:
Pushed a fix for the `scrollTop` crash, which was related to us trying to access `document.body` before it was available.

RE: that `data-brackets-id...not found` error, I'll just remove those logs.  They aren't useful.

humphd commented:
@flukeout pushed a fix for those "data-brackets-id" errors, shouldn't happen now.

flukeout commented:
Ok, will pull latest and continue testing.

flukeout commented:
<img> tag still works in my preview after moving the image file to a folder within the root. I guess maybe it's caching the URL to the image?

Check the code here for the ``jam.jpg`` image.
![image](https://user-images.githubusercontent.com/25212/27097359-d2e9eb1e-5028-11e7-9c73-a7e7b08c27c3.png)


Refreshing the tab with the editor causes this image to correctly not work in the preview.

flukeout commented:
Autocomplete offers invalid suggestions when you type in a ``/`` as the starting character. Neither of the two options in the dropdown below work when selected...

![image](https://user-images.githubusercontent.com/25212/27097582-81c677ec-5029-11e7-98c5-64ee0e894e8d.png)

flukeout commented:
The leaking thing... happens even if you just add a bunch of new lines to the HTML page. I guess that causes the page to refresh a bunch (just like hitting the reload button). This doesn't happen in staging though...

![enter-leak](https://user-images.githubusercontent.com/25212/27098307-f7f47098-502b-11e7-9550-703ef8becd62.gif)

flukeout commented:
Here's another weird thing...

![image-rename-chef](https://user-images.githubusercontent.com/25212/27098645-2e977306-502d-11e7-897d-0453623bf4b4.gif)

So what's happening here is that "slam.jpg" is an old image I had in the project. Renaming an existing image to that name shows us the old one (even though it's no longer in the file tree) and the renaming back to the chef, brings the chef back. Wacky!

flukeout commented:
Looks like when i apply an image filter then re-open the image, the filter is not applied. Refreshign editor makes the applied changes take hold though.

![img-filter](https://user-images.githubusercontent.com/25212/27098977-2ed8c40e-502e-11e7-8004-90041d543a8a.gif)

flukeout commented:
Leak info is here...

![image](https://user-images.githubusercontent.com/25212/27099547-2880e738-5030-11e7-8e03-1e10fb03e39c.png)

humphd commented:
@flukeout I've pushed a commit that *should* deal with things like moving files and such.  If you want to test again, let me know if that works.  @gideonthomas in this change, see 76e5a76, I am wiping the cache when we call `FileSystemCache.refresh()` since it was leaving stale content in there before, which would hang around and show up in odd places (e.g., move an image, and it still shows in the preview).  See if you think this is OK.  It's a bit expensive, but should only be necessary when we do sweeping changes to the fs, so it won't happen too often.

I'll deal with the other stuff tomorrow.

humphd commented:
@flukeout fixed those event listener leaks. Other than the `<a>` links to internal files sometimes not working, which I can't reproduce, I think I've solved everything you found above.  Want to retest?

flukeout commented:
Def. Will test in AM tomorrow first-thing.

flukeout commented:
Okay, got a new one...

**Steps**
* Prepare two different images on your computer, but give them the same name
  * blah.jpg
  * blah.jpg
* Drag the first blah.jpg into your project, everything is fine.
* Drag the second blah.jpg into your project
* Caman will throw an error (seen below)
* Click on the image in the File view
* It's still the original image
* Now, reload the editor
* Click on the image again
* It's the **second** image... whoaaaaaa

Caman bug...
![image](https://user-images.githubusercontent.com/25212/27143814-7a0cfef2-50e4-11e7-8af9-7564f9bd8255.png)

cc @humphd

flukeout commented:
Another thing, also related to links. If you have...

``<a href="">test</a>``

The console goes crazy when you click it.

![image](https://user-images.githubusercontent.com/25212/27144173-b88b2284-50e5-11e7-8568-e0c728b52256.png)

humphd commented:
Fixed all the bugs listed above, with the exception of a couple that are not related to this PR and I've filed separately.  I've also rebased to get the new image resize code, the dual console logging, etc.

flukeout commented:
Small nit, any way to rewrite messages in our own console so they don't reference the crazy ``localhost`` paths?

![image](https://user-images.githubusercontent.com/25212/27153671-3827436c-5107-11e7-8087-f6212e5a3c0c.png)

We could file as a follow up maybe.

flukeout commented:
@humphd Draggin in a new image with the same name as an existing image works better now...

* The preview updates like it should if the image is included on the page
* The Image Viewer doesn't update however, and updates only when you click away from it and click on it again

flukeout commented:
Here is something weird, in your project, add a link to an image file that's in the project...

``<a href="buster.jpg">Buster</a>``

When you click it in the preview, the Preview doesn't change, but the editor switches to show the image file.

humphd commented:
RE: https://github.com/mozilla/brackets/pull/725#issuecomment-308551979 and showing `localhost` URLs, agree, we need to rewrite URLs to filepaths before our console shows them.  That's worth filing as a separate thing, since it also would happen with Blob URLs.

RE: https://github.com/mozilla/brackets/pull/725#issuecomment-308552946 and image viewer not updating, I filed that at https://github.com/mozilla/thimble.mozilla.org/issues/2308, it's unrelated to this PR, but something we should explore fixing.

RE: https://github.com/mozilla/brackets/pull/725#issuecomment-308555998 and editor showing the image, that's expected behaviour, since we navigate to any file in our filesystem if you link to it.  We could blacklist images, but I think it's nice that it does that, so you can find the file again, see colour info, size info, etc.

flukeout commented:
OK, I'm okay with this stuff then - we can address all of these cases in subsequent issues.

flukeout commented:
Fixes:

* https://github.com/mozilla/thimble.mozilla.org/issues/2290

gideonthomas commented on src/extensions/default/bramble/lib/XHRManager.js:
```diff
@@ -0,0 +1,61 @@
+        return "";
+    }
+
+    function getNavigationPath(message) {
```
How is this different than the next function `isXHRRequest`? Seems to return the same output, no?

gideonthomas replied:
Doesn't look like this function is used anywhere

gideonthomas commented on src/extensions/default/bramble/main.js:
```diff
@@ -28,9 +28,7 @@ define(function (require, exports, module) {
-        BlobUtils            = brackets.getModule("filesystem/impls/filer/BlobUtils"),
-        XHRHandler           = require("lib/xhr/XHRHandler"),
-        Theme                = require("lib/Theme"),
+        UrlCache            = brackets.getModule("filesystem/impls/filer/UrlCache"),
```
nit: alignment

gideonthomas commented on src/filesystem/impls/filer/FileSystemCache.js:
```diff
@@ -7,46 +7,42 @@ define(function (require, exports, module) {
                 var mime = Content.mimeFromExt(Path.extname(decodedFilename));
-                var url = BlobUtils.createURL(filename, data, mime);
-                callback(null, url);
+                UrlCache.createURL(filename, data, mime, function(err, url) {
```
nit:
```js
UrlCache.createURL(filename, data, mime, callback);
```

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+    BlobUrlProvider.prototype.constructor = BlobUrlProvider;
+
+    BlobUrlProvider.prototype.init = function(callback) {
+        _.defer(callback);
```
why do we wait to call this vs. just calling callback directly?

humphd replied:
We could; but it gives the two providers more similar startup characteristics, by making this a true async call.

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+    };
+    BlobUrlProvider.prototype.clear = function(callback) {
+        var urls = Object.keys(this.urls);
+        urls.forEach(function(url) {
```
nit:
```js
Object.keys(this.urls).forEach(URL.revokeObjectURL);
```

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+
+        // We use cache URLs like https://<origin>/dist/vfs/project/root/filename.ext
+        var prefix = brackets.env === "production" ? "/dist" : "/src";
+        this.generateVFSUrlForPath = function(path) {
```
any reason this isn't on the prototype?

humphd replied:
There was a reason, but I've refactored it.  Moved to proto.

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+        UrlProviderBase.call(this);
+
+        // We use cache URLs like https://<origin>/dist/vfs/project/root/filename.ext
+        var prefix = brackets.env === "production" ? "/dist" : "/src";
```
Hmm, so this will break if we map `https://mozillathimblelivepreview.net/bramble/staging/dist` to a new origin domain like `https://bramble.net`?

humphd replied:
I can fix this so we don't need to know anything about where it's hosted.

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+        this.clear(callback);
+    };
+    CacheStorageUrlProvider.prototype.createURL = function(filename, blob, type, callback) {
+        var response = new Response(blob, {
```
I'm curious about how this is passing linting since it isn't defined in the `global` comment at the beginning of the file

humphd replied:
I checked and we are linting this file.

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+            if(pathPart === path || pathPart.indexOf(path + "/") === 0) {
+                removed.push(pathPart);
+
+                delete self.urls[pathPart];
```
this won't delete the paths that are prefixed with `pathPart`, but it should?

humphd replied:
That is likely a bug in both of them.  I'll file it as a follow up to fix later vs. blocking this on it.

gideonthomas commented on src/filesystem/impls/filer/UrlCache.js:
```diff
@@ -0,0 +1,381 @@
+                 callback(err);
+             });
+    };
+    CacheStorageUrlProvider.prototype.rename = function(oldPath, newPath, callback) {
```
This won't work for folder renames though?

humphd replied:
Same problem for Blob, no?  I'll file a follow up.

humphd replied:
Filed https://github.com/mozilla/thimble.mozilla.org/issues/2314 for later.

humphd commented on src/LiveDevelopment/MultiBrowserImpl/documents/LiveCSSDocument.js:
```diff
@@ -64,6 +78,8 @@ define(function LiveCSSDocumentModule(require, exports, module) {
     var LiveCSSDocument = function LiveCSSDocument(protocol, urlResolver, doc, editor, roots) {
         LiveDocument.apply(this, arguments);
 
+        this.roots = normalizeRoots(this.roots);
```
I want to leave it as `this.roots` in case the base constructor messes with it in any way, after the call on 79.

gideonthomas replied:
oh I see, missed the constructor call above!

humphd commented on src/extensions/default/bramble/lib/PostMessageTransport.js:
```diff
@@ -207,10 +211,10 @@ define(function (require, exports, module) {
-            "<script>\n" + XHRShim + "</script>\n" +
-            MouseManager.getRemoteScript(escapedPath) +
+            XHRManager.getRemoteScript() +
+            MouseManager.getRemoteScript(currentPath) +
```
Good catch, this was reverted accidentally in a rebase.  See why I need/want you reviewing my crappy code?  :)

humphd commented on src/extensions/default/bramble/lib/UI.js:
```diff
@@ -249,13 +245,6 @@ define(function (require, exports, module) {
         }
 
-        // Switch the icon
-        $("#mobileViewButton").removeClass("desktopButton");
```
This was code we did at the very, very beginning at CDOT, when we wanted to allow you to do various things from the status bar of the editor, before we had any UI in Thimble to control things.  It's dead now.

humphd commented:
@gideonthomas, @flukeout: this has been rebased onto master and squashed to a single commit.  One of us can land this after we push staging to prod, it's all set to go!

humphd commented:
This stuff has now landed on staging if people want to play with it:

https://bramble.mofostaging.net

Please file bugs, as I'm sure this will have broken things, given how much it changes internally.
